controllers/agentcluster_controller.go

```diff
@@ -84,17 +90,63 @@ func (r *AgentClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
+	patchHelper, err := patch.NewHelper(agentCluster, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	defer func() {
+		if err = patchHelper.Patch(ctx, agentCluster); err != nil {
+			log.WithError(err).Errorf("failed patching AgentCluster")
+		}
+		log.Info("AgentCluster Reconcile ended")
+	}()
+
+	// AgentCluster is being deleted
+	if !agentCluster.DeletionTimestamp.IsZero() {
+		if err = r.handleDeletion(ctx, agentCluster); err != nil {
+			log.WithError(err).Errorf("failed to remove AgentCluster")
+			return ctrl.Result{}, err
+		}
+	}
+
+	if paused := agentCluster.Annotations[clusterv1.PausedAnnotation]; paused == "true" {
+		log.Info("Skipping reconcile of AgentCluster as it's paused, but orphan its resources")
+		if err := r.orphanResources(ctx, agentCluster); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	if !controllerutil.ContainsFinalizer(agentCluster, agentClusterFinalizer) {
```

Comment:
I think we should move this up. Typically the logic should be right near where we check if the object is being deleted.

Reply:
Good point, thanks! Modified.